### PR TITLE
finalized api challenge

### PR DESCRIPTION
**Pull request by gmanchon**
add challenge instructions in the README
add a Makefile in order to ease the use of Docker
update the Dockerfile in order to copy model.joblib and use $PORT
add a rudimental trainer used in order to generate the model.joblib
add a model.joblib
fix the required packages versions in requirements.txt

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,17 +1,227 @@
...
-```bash
-pip install -e .
+### Duplicate the repository for the API challenge
+
+**📝 Let's duplicate the repository of the API challenge.**
+
+Go to https://github.com/lewagon/data-certification-api:
+- Click on `Use this template`
+- Enter the repository name `data-certification-api`
+- Set as **Public**
```
Comment (krokrob) [inline README.md]:
```suggestion
- Set it as **Public**
```

File: README.md [hunk 2]
```diff
@@ -1,17 +1,227 @@
...
+- Click on `Create repository from template`
+- Click on `Code`
+- Select `SSH`
+- Copy the SSH URL of the repository (the format is `git@github.com:YOUR_NAME/data-certification-api.git`)
```
Comment (krokrob) [inline README.md]:
```suggestion
- Copy the SSH URL of the repository (the format is `git@github.com:YOUR_GITHUB_NICKNAME/data-certification-api.git`)
```

File: README.md [hunk 3]
```diff
@@ -1,17 +1,227 @@
...
+│   │   └── __init__.py
+│   └── utils.py
+├── notebooks
+├── requirements.txt
+├── scripts
+│   └── exampack-run
+└── setup.py
+```
+
+You can now open you favorite text editor and proceed with the challenges.
```
Comment (krokrob) [inline README.md]:
```suggestion
Open your favourite text editor and proceed with the challenge.
```

File: README.md [hunk 4]
```diff
@@ -1,17 +1,227 @@
...
+
+## API challenge
+
+**📝 In this challenge, you are provided with a trained model saved as `model.joblib`. The goal is to create an API that will predict the popularity of a song based on its other features.**
+
+👉 In this challenge you will only need to edit the code of the API in `api/app.py` 🚨
+
+👉 The package versions listed in `requirements.txt` should work out of the box will the pipeline saved in `model.joblib`
```
Comment (krokrob) [inline README.md]:
```suggestion
👉 The package versions listed in `requirements.txt` should work out of the box with the pipelined model saved in `model.joblib`
```

File: README.md [hunk 5]
```diff
@@ -1,17 +1,227 @@
...
+
+### Install the required packages
+
+**📝 Let's install the required packages.**
+
+👉 The `requirements.txt` file lists the exact version of the packages required in order to be able to load the trained pipeline that we provide
```
Comment (krokrob) [inline README.md]:
```suggestion
👉 The `requirements.txt` file lists the exact version of the packages required in order to be able to load the pipelined model that we provide
```

File: README.md [hunk 6]
```diff
@@ -1,17 +1,227 @@
...
+If you encounter an version conflict while installing the packages, you will need to work in a new virtual environment:
+
+👉 Only execute this commands if you encounter an issue while installing the packages 🚨
+
+``` bash
+pyenv install 3.8.6
+pyenv virtualenv 3.8.6 certif
+pyenv local certif
+pip install -r requirements.txt
+```
```
Comment (krokrob) [inline README.md]: make it collapsable please

File: README.md [hunk 7]
```diff
@@ -1,17 +1,227 @@
...
+
+## API endpoint
+
+**📝 Let's begin by creating a prediction endpoint.**
```
Comment (krokrob) [inline README.md]: to be removed

File: README.md [hunk 8]
```diff
@@ -1,17 +1,227 @@
...
+👉 Once `X_pred` is built, you may `print` it and have a look at the output of the `uvicorn` command in order to make sure that it is correctly created
+
+### Make a prediction
+
+**📝 Now that you have a trained pipeline and a `DataFrame` containing one observation, let's make a prediction and return it.**
+
+👉 Remember that `FastAPI` can only handle simple data types as returned `json` values
+
+Test your API using http://localhost:8000/docs and make sure that it works correctly before proceeding to put it in production.
+
```
Comment (krokrob) [inline README.md]: ### API specification

**Predict the popularity of a Spotify song**

`GET /Predict`

| Parameters | Description |
|---|---|
| acousticness | floating number for acousticness |
| danceability | todo |
| duration_ms | todo |
| energy | todo |
| explicit | todo |
| id | todo |
| instrumentalness | todo |
...

Returns a dictionary with the `artist`, the `name` of the song and predicted `popularity` as an integer.

Example request:

```
/predict?acousticness=0.654&danceability=0.499&duration_ms=219827&energy=0.19&explicit=0&id=0B6BeEUd6UwFlbsHMQKjob&instrumentalness=0.00409&key=7&liveness=0.0898&loudness=-16.435&mode=1&name=Back%20in%20the%20Goodle%20Days&release_date=1971&speechiness=0.0454&tempo=149.46&valence=0.43&artist=John%20Hartford
```

Example response:
```json
{
  "artist": "John Hartford",
  "name": "Back in the Goodle Days",
  "popularity": 28
}
  
```

File: README.md [hunk 9]
```diff
@@ -1,17 +1,227 @@
...
+## API in production
+
+**📝 Push your API to production on the hosting service of your choice.**
+
+👉 Remember to test your API locally and make an actual prediction before pushing your code to production in order to save time
+
+👉 You may test your API with the following parameters (change the domain and port with your own):
+
+```
+http://localhost:8000/predict?acousticness=0.654&danceability=0.499&duration_ms=219827&energy=0.19&explicit=0&id=0B6BeEUd6UwFlbsHMQKjob&instrumentalness=0.00409&key=7&liveness=0.0898&loudness=-16.435&mode=1&name=Back%20in%20the%20Goodle%20Days&release_date=1971&speechiness=0.0454&tempo=149.46&valence=0.43&artist=John%20Hartford
```
Comment (krokrob) [inline README.md]: To be removed

File: README.md [hunk 10]
```diff
@@ -1,17 +1,227 @@
...
+
+**📝 Now that you have a trained pipeline and a `DataFrame` containing one observation, let's make a prediction and return it.**
+
+👉 Remember that `FastAPI` can only handle simple data types as returned `json` values
+
+Test your API using http://localhost:8000/docs and make sure that it works correctly before proceeding to put it in production.
+
+## API in production
+
+**📝 Push your API to production on the hosting service of your choice.**
```
Comment (krokrob) [inline README.md]: OK but the Makefile is for GCR only!!

File: README.md [hunk 11]
```diff
@@ -1,17 +1,227 @@
...
+
+👉 You may test your API with the following parameters (change the domain and port with your own):
+
+```
+http://localhost:8000/predict?acousticness=0.654&danceability=0.499&duration_ms=219827&energy=0.19&explicit=0&id=0B6BeEUd6UwFlbsHMQKjob&instrumentalness=0.00409&key=7&liveness=0.0898&loudness=-16.435&mode=1&name=Back%20in%20the%20Goodle%20Days&release_date=1971&speechiness=0.0454&tempo=149.46&valence=0.43&artist=John%20Hartford
+```
+
+💡 With a configured GCP account, the directives in the `Makefile` should be enough to push your API to production
+
+👉 Remember to update the `GCP_PROJECT_ID` in the `Makefile` 🚨
```
Comment (krokrob) [inline README.md]: Once you had change your `GCP_PROJECT_ID` in the `Makefile`, run the following command to build and deploy your containerized API to GC Container Registry and finally GCRun.

File: exampack/predictor.py [hunk 12]
```diff
@@ -1,2 +1,63 @@
...
+
+        pipeline = Pipeline(steps=[
+            ("column_transformer", column_transformer),
+            ("model", LinearRegression())
+        ])
+
+        return pipeline
+
+    def train(self):
+
```
Comment (krokrob) [inline exampack/predictor.py]: Here we could detail the specificity of the pipeline: what is X, what is y to avoid a long list of info in the README

**Review by krokrob — COMMENTED** [review]
We need to size down the README, I wrote some suggestions, happy to discuss

File: README.md (hunk 11 above)
Comment (gmanchon) [reply README.md]: I added the instructions to test locally then push and deploy 🙏

File: README.md (hunk 10 above)
Comment (gmanchon) [reply README.md]: I added the instructions for GCP in a collapsed paragraph 🙏

**Comment by gmanchon** [comment]
@krokrob thanks for the review 🙏

**Pull request merged**